File: experimental_pure_python_port/working_version2_start1_removedComments/aprodw.py
# Python modules
from __future__ import division
import pdb

# 3rd party modules
import numpy as np


# Our modules



def aprodw(transa, ndp, m, n, z, y, z1, y1, lambda_, trlambda):
    """
    Computes y=TH*T*z where lambda contains the eigenvalues of the circulant
    matrix used for T and trlambda contains the eigenvalues of the circulant
    matrix used for TH.

    parameter (zeroc=(0.0d0,0.0d0))
    integer         m,n,ndp
    complex*16      z(*),y(*),z1(ndp),y1(ndp),lambda(*),trlambda(*)
    character*1     transa
    
    """
    transa = transa.lower()
    if transa == 't':
        z1[:m] = z[:m]

        z1[m:ndp] *= 0j

        z1_copy, y1_copy = ztmultz(trlambda, ndp, z1, y1)

        z1[:ndp] = z1_copy[:ndp]
        y1[:ndp] = y1_copy[:ndp]
        y[:n] = y1[:n]

    else:
        z1[:n] = z[:n]

        z1[n:ndp] *= 0j

        z1_copy, y1_copy = ztmultz(lambda_, ndp, z1, y1)
        z1[:ndp] = z1_copy[:ndp]
        y1[:ndp] = y1_copy[:ndp]

        y[:m] = y1[:m]

    return z, y, z1, y1




def ztmultz(lambda_, ndp, z, y):
    """
    Multiply a Toeplitz matrix by a vector with the help of FFT

    integer         ndp
    double complex  lambda(*), z(*), y(*)

    """
    y = np.fft.fft(z)

    z = y * lambda_

    y = np.fft.ifft(z)

    # The FFTW doc says, "FFTW computes an unnormalized transform, in that 
    # there is no coefficient in front of the summation in the DFT. In other 
    # words, applying the forward and then the backward transform will multiply 
    # the input by n."
    # http://www.fftw.org/faq/section3.html#whyscaled
    # Apparently numpy's FFT is normalized, so in order to get the same 
    # results as FFTW computes, we have to multiply by n a.k.a. ndp.
    y *= ndp

    return z, y

File: hlsvdpro_v1/src/aprodw.f

      subroutine aprodw(transa,ndp,m,n,z,y,z1,y1,lambda,
     c                  trlambda,planF,planB)


c
c     Computes y=TH*T*z where lambda contains the eigenvalues of the circulant
c     matrix used for T and trlambda contains the eigenvalues of the circulant
c     matrix used for TH.
c

      parameter (zeroc=(0.0d0,0.0d0))
      integer m,n,ndp
      complex*16 z(*),y(*),lambda(*),trlambda(*),z1(ndp),y1(ndp)

      character*1 transa
      integer i,planF,planB
      logical lsame
      external lsame 

      if (lsame(transa,'t')) then
        do i=1,m
           z1(i)=z(i)
        end do
        do i=m+1,ndp
           z1(i)=zeroc
        end do

        call ztmultz(trlambda(1),ndp,z1(1),y1(1),planF,planB)
        
        do i=1,n
           y(i)=y1(i)
        end do
      else
        do i=1,n
           z1(i)=z(i)
        end do
        do i=n+1,ndp
           z1(i)=zeroc
        end do

        call ztmultz(lambda(1),ndp,z1(1),y1(1),planF,planB)
        do i=1,m
           y(i)=y1(i)
        end do

      end if

      end



      subroutine ztmultz(lambda,ndp,z,y,planF,planB)
c
c Multiplicerar en toepliz-matris med en vektor med hjalp av fft
c
      integer m,n, ii

      integer i,ndp,planF,planB
      double complex lambda(*), z(*), y(*)
C      double complex fz(m+n),temp


      
c       do ii = 1, m + n
c          fz(ii)=(0.0d0,0.0d0)
c       end do
c      call zcopy(n,z(1),1,fz(1),1)
c      call dcfftf(m+n,fz(1),wsave(1))

c      y=FFT(z)

      call dfftw_execute_dft(planF, z, y)   


c
      do ii = 1, ndp
         z(ii)=y(ii)*lambda(ii)
      end do
      call dfftw_execute_dft(planB, z, y)   

c      call dcfftb(m+n,fz(1),wsave(1))
c      call zcopy(m,fz(1),1,y(1),1)
cC      temp=cmplx(1.0d0/dble(m+n),0.0d0,kind(1.0d0))
c      temp=cmplx(1.0d0/dble(m+n),0.0d0)
c      call zscal(m,temp,y(1),1)
      end

